doc/developer/design/20230802_webhook_idempotency.md

```diff
@@ -0,0 +1,403 @@
+# Webhook idempotency
+
+- Associated:
+    - [Webhook design
+      doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20230615_webhook_source.md)
+
+## Context
+
+To make webhook sources offer sane durability semantics, we need to make their
+"inserts" (i.e. `POST`s) idempotent––for example, Segment will try to make
+webhook `POST`s for multiple hours until they receive a terminal status (i.e. I
+believe they retry on 5xx errors).
+
+## Goals
+
+Propose a solution to make webhook sources support idempotency on a configurable
+timeline (e.g. 24 hours). This style of idempotency is supported by other
+webhook integrations, such as
+[SaaSquatch](https://docs.saasquatch.com/developer/event-idempotency#event-idempotency).
+
+Note that this form of idempotency should also be durable, e.g. idempotent for
+24 hours independent of Materialize (or any of its components) restarting.
+
+## Non-Goals
+
+- Make all requests idempotent for all time (requires infinite memory)
+- Ensure that webhooks are usable within a certain price point
+- Ensure that webhooks achieve some specific throughput/latency. We want these
+  to be fast enough to be useful but introducing additional disk writes cannot
+  be a deal breaker. 
+
+## Overview
+
+The simplest means by which we can offer idempotency for webhooks is using
+`UPSERT` envelope semantics for an idempotency key included in the requests'
+headers or bodies.
+
+To simplify the implementation, we plan to use the exact same machinery that the
+`UPSERT` envelope uses, which means that webhook sources will need to run on
+`clusterd` (they currently run on `environmentd`).
+
+## Detailed description
+
+### Semantic overview
+
+#### SQL
+
+When creating a source, just like `CHECK`, we will have a new `IDEMPOTENT`
+clause that will give us an expression through which we can extract the
+idempotency key from the message.
```

Comment:
I'd been envisioning that we would use the `ENVELOPE UPSERT (KEY (...))` syntax that we invented for `SUBSCRIBE`: https://materialize.com/docs/sql/subscribe/#envelope-upsert. IMO, that's the syntax we should have always had, for both sources and sinks! Fully rendered it would look like:

```
CREATE SOURCE ... FROM WEBHOOK
ENVELOPE UPSERT (KEY (body->>'key'))
```

But possibly we want to not reuse the `ENVELOPE UPSERT` syntax, since it's not like we'd ever support e.g. `ENVELOPE DEBEZIUM`.